### ISSUE TITLE
sp_BlitzFirst - Server Uptime vs Query Runtime

### ISSUE THREAD

**Issue by erikdarlingdata** [BrentOzarULTD/SQL-Server-First-Responder-Kit#2913]
**Is your feature request related to a problem? Please describe.**
When looking at wait stats, the relationship between total wait time and server uptime can be useful to highlight when contention is constant. 

However, for servers that aren't constantly slammed or sometimes look bored, it can be harder to justify certain things like adding hardware. For that, you really need to catch contention in the act.

**Describe the solution you'd like**
I'd like to take a stab at figuring how much query runtime there is since the server has been started.. To do this, we need to sum `wait_time` from `dm_os_wait_stats` and `total_cpu_usage` from `dm_os_schedulers` for visible online cores only. I also need to rule out a number of waits, and that list can come from the waits temp table we already populate.

A starting point for this query [is here](https://gist.github.com/erikdarlingdata/c96f6093a3eb8bbb3cc3042f0d3599ae), though I'm on the fence about `SLEEP_TASK` being ignored. It can crop up during hash spills and other unpleasantness.

**Describe alternatives you've considered**
Not day drinking

**Are you ready to build the code for the feature?**
Obviously, if people out there think this is a feasible and useful metric for comparison.

Thanks to Sean Gallardy and Paul White for the sanity checks on the starting point query, of course.

**Comment by erikdarlingdata** [comment]
I've made a couple updates to the script after testing to
* Using a view for portability/testing
* Remove signal wait time, since it's already included in wait time (thanks Paul)
* Format results a little more nicely

I've been testing this with various queries, and the results seem to line up accurately. I was thrown off a bit at first by testing with parallel queries, but since "waits in parallel plans accumulate for each thread, it makes sense to compare with total thread time"

**Comment by BrentOzar** [comment]
This is a brilliant, brilliant idea.

**Comment by erikdarlingdata** [comment]
I have this working for the `@SinceStartup` and `@Seconds` parameters, but I haven't implemented it in the logging portions because I don't want to introduce breaking changes until if/when things are finalized. There's a pull request available if anyone wants to test/tinker.

![image](https://user-images.githubusercontent.com/2136037/120495727-f8159800-c38a-11eb-90ed-9b4a5666cb6c.png)

**Comment by erikdarlingdata** [comment]
@BrentOzar no rush on this, but I bumped the ignore list I had up against the ignoreables in BlitzFirst, and found these differences:

![image](https://user-images.githubusercontent.com/2136037/120524430-0e305200-c3a5-11eb-8cef-e86275633e67.png)

I do agree with `PREEMPTIVE_OS_FLUSHFILEBUFFERS`, `CHKPT` and `SLEEP_BPOOL_FLUSH` not being generally ignoreable -- I can tweak the query for those -- but the rest I think we may want to re-designate.

**Comment by erikdarlingdata** [comment]
Noticed a potential issue with the starting query where it's not touching `sys.dm_os_waiting_tasks` to get _current_ waits. It's only hitting `dm_*_wait_stats` for _past_ waits. Will have to tweak things a little bit to make sure they match a little better.

**Comment by erikdarlingdata** [comment]
After messing with this for a while, I realized I was going about it all wrong. Instead of making this a totally separate query, all I had to do was grab the additional CPU timing in the regular wait stats query, and then update each pass to add the sum of all wait times to the cpu time. That seems to get correct number for both total uptime and sampled times.

Total uptime:
![image](https://user-images.githubusercontent.com/2136037/120572930-cb916880-c3ea-11eb-88fc-d5a19f85775e.png)

Sampled run:
![image](https://user-images.githubusercontent.com/2136037/120573082-085d5f80-c3eb-11eb-8da2-e29dd2f1717c.png)

I'm about cooked for tonight working on this, but I'm going to do some more checking in the morning to make sure I'm not missing anything

**Comment by erikdarlingdata** [comment]
alright, one more tweak to this because [non-optimistic isolation levels are garbage](https://youtu.be/zBbzGF9lG1U). results were inconsistent in the thread time column and i needed to add an update to fix them.

full explanation: the non-correlated scalar subquery to get cpu time was implemented in the execution plan as a nested loops join so it was possible to get different results each time the loop executed. to fix this, i use the same cte to add wait time to thread time to use the max thread time instead of the potentially inconsistent per-row results.

**Comment by BrentOzar** [comment]
tl;dr - I love this, works great.

Long story - at first glance, I thought it was underestimating thread time because it wouldn't include how much time queries were actively consuming CPU, and not waiting on anything. (The good ol' "CPU/Memory" number from a well-known tool.)

To test that, I ran this:

```
USE StackOverflow;
GO
SET STATISTICS TIME ON;
GO
/* First time gets it in memory */
SELECT COUNT(*) FROM dbo.Users WITH (INDEX = 1);
GO
/* Then look at stats time & the actual plan for this: */
SELECT COUNT(*) FROM dbo.Users WITH (INDEX = 1) OPTION (MAXDOP 1);
GO

```

Once the data's up in cache, it used hundreds of milliseconds of CPU time each time it runs, but if you look at the actual plan, SOS_SCHEDULER_YIELD wait times are very small (like 10 or less milliseconds) because there's no competition for CPU.

To load test the problem, I ran that in SQLQueryStress executing that query with just 4 threads (to keep CPU contention low, but still burn a lot of CPU.) The server's CPU was high:

<img width="664" alt="Screen Shot 2021-06-06 at 7 32 17 AM" src="https://user-images.githubusercontent.com/245462/120916508-86df2900-c699-11eb-9dab-30cd89e09a85.png">

But there wasn't much CPU wait time because no other queries were running. Here's what sp_BlitzFirst @ExpertMode = 1, @Seconds = 30 reported:
<img width="953" alt="Screen Shot 2021-06-06 at 7 32 39 AM" src="https://user-images.githubusercontent.com/245462/120916517-96f70880-c699-11eb-8188-f818cf18c747.png">

GORGEOUS. SOS_SCHEDULER_YIELD wait times are low (1.3 seconds), but thread time still reports as 117.6 seconds because a lot of queries were using CPU during that time span.

Similarly, @SinceStartup = 1 does a great job too:
<img width="947" alt="Screen Shot 2021-06-06 at 7 35 10 AM" src="https://user-images.githubusercontent.com/245462/120916550-c574e380-c699-11eb-91c1-cb68daa61371.png">

Great work!

**Comment by BrentOzar** [comment]
Good news and bad news. The good news is, it's accurate because sys.dm_os_schedulers added a total_cpu_usage_ms column. (I learned this just now by reading your code, hahaha. I was like, how's he getting total CPU usage for queries that have already run in the past, without hitting the plan cache and assuming some data loss?)

The bad news is, it's only 2016 and newer. 

We'll just need to frame this up with dynamic SQL:

* When that total_cpu_usage_ms field exists, use it as-is, and the numbers are perfectly accurate
* When that field doesn't exist, we could sum up the total_worker_time from the plan cache and use that as a substitute, and make a note in the documentation that it's underestimated
* Or, when that field doesn't exist, we could simply return 0s in that column (to make it clear that we don't have an answer) - I'm open to that too.

**Comment by erikdarlingdata** [comment]
Man, I got so excited about doing this I totally forgot to check for backwards compatibility. Sorry about that.

I'm cool with showing the plan cache sum of total worker time, but I'd also be fine with making everything extra dynamic and not showing the column where it isn't supported.

If we're going to look at the plan cache, we may want look at how old it is compared to server uptime and surface a warning or something.

```
SELECT
    sm = SUM(total_worker_time),
    mn = MIN(creation_time),
    up = 
    (
        SELECT
            dosi.sqlserver_start_time
        FROM sys.dm_os_sys_info AS dosi
    )
FROM sys.dm_exec_query_stats;
```

I've been all over the place today and haven't had much time to really think about what the better option is. I'm open to whatever though.

**Comment by erikdarlingdata** [comment]
Getting back into this, and trying to decide on workarounds, I don't think the plan cache thing is going to work. For some reason, it's *way* higher than the CPU THREAD time. 

Here are what I get from [these queries](https://gist.github.com/erikdarlingdata/bc2d3d08683381bb46f7672461b11997): 

![image](https://user-images.githubusercontent.com/2136037/121778868-f86e1a00-cb66-11eb-91dc-ac4aa5583a63.png)

That seems... Odd. So odd I don't even know what to do with it. My instance has only been up for a couple days, too. That worker time in the plan cache is nearly 8X what's recorded in `total_cpu_usage_ms` is bonkers. 

Thinking that maybe there was some accounting oddities about, I tried this:

```
SELECT
    from_cache = 
        FORMAT
        (
            SUM(total_worker_time 
                    / execution_count), 
            N'N0'
        )
FROM sys.dm_exec_query_stats;
```

That gives me a closer number, but it still doesn't feel right:
![image](https://user-images.githubusercontent.com/2136037/121779138-37509f80-cb68-11eb-8cb3-4ab47cf91c9f.png)

I tried the same thing with `total_dop`, but that column isn't available < 2016 anyway.

Rather than use a mind-boggler drop-in, I'd rather just return 0 and not produce a metric I can't explain.

**Comment by erikdarlingdata** [comment]
Ah, I was being dumb when I looked at this last. `total_worker_time` is in microseconds -- dunno why that skipped my mind. I think we'll be able to use it as a drop-in number instead of just reporting 0.

**Comment by BrentOzar** [comment]
I've banged around on this for about an hour on SQL Server 2019, and I can't get it to break. Looks good! Merging into the dev branch now, and going to use it in this week's Mastering Server Tuning class to see if I spot issues during that.
